link-grammar/parse/prune.c

```diff
@@ -129,99 +129,59 @@ struct prune_context_s
 };
 
 /*
-  FIXME: These comments are old and are not relevant in their
-  current form. Convert them to relevant comments.
-
-  The algorithms in this file prune disjuncts from the disjunct list
-  of the sentence that can be eliminated by a simple checks.  The first
-  check works as follows:
-
-  A series of passes are made through the sentence, alternating
-  left-to-right and right-to-left.  Consider the left-to-right pass (the
-  other is symmetric).  A set S of connectors is maintained (initialized
-  to be empty).  Now the disjuncts of the current word are processed.
-  If a given disjunct's left pointing connectors have the property that
-  at least one of them has no connector in S to which it can be matched,
-  then that disjunct is deleted. Now the set S is augmented by the right
-  connectors of the remaining disjuncts of that word.  This completes
-  one word.  The process continues through the words from left to right.
-  Alternate passes are made until no disjunct is deleted.
-
-  It worries me a little that if there are some really huge disjuncts lists,
-  then this process will probably do nothing.  (This fear turns out to be
-  unfounded.)
-
-  Notes:  Power pruning will not work if applied before generating the
-  "and" disjuncts.  This is because certain of it's tricks don't work.
-  Think about this, and finish this note later....
-  Also, currently I use the standard connector match procedure instead
-  of the pruning one, since I know power pruning will not be used before
-  and generation.  Replace this to allow power pruning to work before
-  generating and disjuncts.
-
-  Currently it seems that normal pruning, power pruning, and generation,
-  pruning, and power pruning (after "and" generation) and parsing take
-  about the same amount of time.  This is why doing power pruning before
-  "and" generation might be a very good idea.
-
-  New idea:  Suppose all the disjuncts of a word have a connector of type
-  c pointing to the right.  And further, suppose that there is exactly one
-  word to its right containing that type of connector pointing to the left.
-  Then all the other disjuncts on the latter word can be deleted.
-  (This situation is created by the processing of "either...or", and by
-  the extra disjuncts added to a "," neighboring a conjunction.)
-
+* Here is what you've been waiting for: POWER-PRUNE
+*
+* The kinds of constraints it checks for are the following:
+*
+*  1) successive connectors on the same disjunct have to go to
+*     nearer and nearer words.
+*
+*  2) two deep connectors cannot attach to each other
+*     (A connectors is deep if it is not the first in its list; it
+*     is shallow if it is the first in its list; it is deepest if it
+*     is the last on its list.)
```

Comment:
This claim: "two deep connectors cannot attach to each other" seems to be immediately contradicted below: "on two adjacent words, a pair of connectors can be used only if they're the deepest" ... !?  I'll merge this pull req, but this neds fixing.

Reply:
It is not a contradiction, because the deepest connector can be a shallow one, while a deep connector cannot be shallow.

Maybe a clarification can be added:
```
    144 *  3) on two adjacent words, a pair of connectors can be used
    145 *     only if they're the deepest ones on their disjuncts.
    146 *     NOTE: the deepest connector can be shallow, while a
    147 *     deep connector cannot be shallow.

Reply:
I find the clarification even more confusing. So, in 2), I guess it is saying that every deep connector must connect to a shallow one? So that:
```
2) two deep connectors cannot attach to each other; every deep connector must 
    connect to a shallow one.
```

Reply:
I tweaked wording above.

Reply:
The implication from "two deep connectors cannot attach to each other" as a principal exclusion rule is slightly wider:
1. A deep connector can only attache to a shallow one. (As you wrote.)
2. A shallow connector can attach to any connector.

So maybe we can change it to:
```
2) two deep connectors cannot attach to each other; i.e. a deep
connector can only attache to a shallow one, and a shallow 
connector can attache to any connector.
```

Reply:
that works